students/CarolF/module05/MailRoom_3_CFarris.py

```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env Python3
+
+################################
+##Created By: Carol Farris
+##Date: February 13th 2018
+##Goal: MailRoom Part 3 Assignment. 
+##Changes: Moved around code from Mailroom2 version to improve testability (for Mailroom 4).
+##Added try catch exception to while loop managing selection at bottom. Couldn't find a good   
+##place to add comprehension--Will try in Mailroom 4 after I have refactored code a bit more. 
+################################
+
+
+usersChoice = {1: 'Send Thank You', 2: 'Create Report', 3 : 'Send letters to everyone', 4: 'Quit'}
+donorDict = {'Peter_Piper' : [100.00, 200.00 , 3000.00], 'Lucille_Ball': [100.00,200.00], 'Barak_Obama': [100.00,2000.00,3000.00],
+             'Mary_Poppins': [1000, 2000.5], 'Bill_Gates': [100.0, 2000.0, 1000.0]} 
+#For Mailroom 4, put donorDict into function to create dummy donor list. then pass to needed functions 
+             
+
+
+def findDonorName():
+    """This function identifies the donor information to be used"""
+    donorName = str(input("Please provide the donor name for the thank you or type 'list' to see the list of donors >"))
+
+    while donorName == 'list': #as long as user types list, it will return the donor list
+        for key in donorDict:  print(key) #dictionary version to provide list
+        donorName = str(input("Please enter the donor name to compose the thank you >"))
+
+    return(donorName)
+
+def retrieveDonationAmount(donorName):
+    """requests donor information and will append new contibution to former donor"""
+    newContribution = float(input("please enter new contribution amount....> "))
+    
+    if(donorName in donorDict):
+         newListValue = donorDict[donorName] 
+         donorDict[donorName].append(newContribution)       
+    else:
+        donorDict[donorName] = [newContribution] #turned value into single value list in anticipation of adding contributions in the future
+
+    return(newContribution)
+
+def printLetter(donorName, newContribution):
+    """completes and prints Thank you letter then calls getAction method to request new action"""
+    print("\nDear {},".format(donorName)) #added newlines to improve readability
+    print("Thank you for your generous contribution of ${} ".format(newContribution))
+    print("Your contribution is greatly appreicated and will be used to grow our operations and will reinforce our commitment to innovation.")
+    print("Sincerely,")
+    print("The Entire Team at PyCert.com\n")
+
+
+#still need to call and improve formattting.
+def createReportData():
+    print('{:<15}'.format("Donor Name"), '{:<20}'.format('|  Total Given  |'),
+                 '{:>5}'.format('Number of Gifts  |'),'{:>20}'.format('Average Gift  |')) 
+    print('_'*77)
+    dataForDonationReport = []
+    for key in donorDict:
```

Comment:
Here's another for loop that can be swapped out for a dictionary or list comprehension because they're faster!